medPython #1: embedding, setup and initialization

fcollot opened the pull request:
_(Edited on 18 June following changes to how Python is embedded)_

This PR introduces the first part of the medPython module. A prepackaged embeddable Python is downloaded from the medInria website and added as a resource to the application. The package also contains the Python headers, which means **_Python is no longer a compilation prerequisite_**.
(the packages can be found at https://med.inria.fr/downloads/embedded-python/ but CMake will download them automatically)

Florent2305 commented on src/layers/medPython/core/medPythonCoreAPI.h:
```diff
@@ -0,0 +1,68 @@
+#include <cstddef>
+
+// The following definitions avoid compilation errors when including Python.h
+#undef slots
```
Il serait bon de documenter chaque define et undef.
Sont ils réinitialiser dans un autre fichier ou cela n'a pas d'importance ?

fcollot replied:
`LARGEFILE_SOURCE` était utilisé pour régler un conflit avec VTK dans le code des pipelines. Maintenant que le Python est séparé du reste, cette définition n'est plus nécessaire alors je vais l'enlever.
J'ai documenté la raison des autres define dans le commit que je vais bientôt pusher.

Florent2305 commented on src/layers/medPython/core/medPythonCoreInit.cpp:
```diff
@@ -0,0 +1,145 @@
+
+    if (success)
+    {
+        wchar_t** moduleSearchPaths = (wchar_t**)PyMem_RawMalloc(2 * sizeof(wchar_t*));
```
Attention au magic number 2 devrait être une constante

Florent2305 commented on src/layers/medPython/core/medPythonCoreInit.cpp:
```diff
@@ -0,0 +1,145 @@
+                  && checkStatus(PyConfig_Read(config)) // required to avoid recalculation of module_search_paths during initialization
+                  && checkStatus(PyConfig_SetWideStringList(config, &config->module_search_paths, 2, moduleSearchPaths));
+
+        for (int i = 0; i < 2; i++)
```
si 2devien une constante alor le **for** a du sens sinon deux lignes serait plus simple à comprendre

Florent2305 commented on src/layers/medPython/medPythonInit.cpp:
```diff
@@ -0,0 +1,45 @@
+
+bool setup()
+{
+    bool success = true;
```
proposition de refacto:
```
bool success = core::setup();    
if (success)
{
    QApplication::connect(qApp, &amp;QApplication::aboutToQuit, &amp;teardown);
}
else
{
    core::teardown();
}
return success;
```

Florent2305 commented on src/layers/medPython/core/medPythonCoreUtils.cpp:
```diff
@@ -0,0 +1,40 @@
+QString wideCharToQString(const wchar_t* input)
+{
+    QString result;
+    char* result_char = Py_EncodeLocale(input, nullptr);
```
mettre un warning dans les commentaires pour indiquer un truc du style:
/// Warning Don't use wideCharToQString before Py_PreInitialize

Florent2305 commented on src/layers/medPython/core/medPythonCoreUtils.cpp:
```diff
@@ -0,0 +1,40 @@
+
+wchar_t* qStringToWideChar(QString input)
+{
+    return Py_DecodeLocale(qUtf8Printable(input), nullptr);
```
idem

Florent2305 commented on src/app/medInria/main.cpp:
```diff
@@ -270,6 +271,8 @@ int main(int argc,char* argv[])
 
     forceShow(*mainwindow);
 
+    med::python::setup();
```
où est la déinitialisation de python

fcollot replied:
La fonction `setup` connecte le fonction `teardown` au signal `aboutToQuit`. J'ai suivi les recommandations de Qt:

> We recommend that you connect clean-up code to the aboutToQuit() signal, instead of putting it in your application's main() function. This is because, on some platforms the QApplication::exec() call may not return. For example, on the Windows platform, when the user logs off, the system terminates the process after Qt closes all top-level windows. Hence, there is no guarantee that the application will have time to exit its event loop and execute code at the end of the main() function, after the QApplication::exec() call.

Florent2305 replied:
nice

Florent2305 reviewed (changes requested):
Great Job.
Juste few very small modifications are needed or answer to questions.
With these modifications we will need python and we will embed python all the time.
I think a good evolution is to conditional compilation if python are present and/or a conditional  execution of python. 
I would be pleased to discuss this with you by visio/phone.

Florent2305 commented:
I can try to compile on windows and ubuntu if you want ?

mathildemerle commented:
> @fcollot I can try to compile on windows and ubuntu if you want ?

I can compile also on Ubuntu :)

fcollot commented on src/layers/medPython/CMakeLists.txt:
```diff
@@ -0,0 +1,194 @@
+find_package(dtk REQUIRED)
+include_directories(${dtk_INCLUDE_DIRS})
+
+find_package(Qt5 REQUIRED Core)
```
COMPONENTS is optional if REQUIRED is present, but I don't mind adding it.

Florent2305 replied:
ok

fcollot commented on src/layers/medPython/core/medPythonCore.cpp:
```diff
@@ -0,0 +1,35 @@
+
+bool teardown()
+{
+    return finalizePython() & teardownCoreLibrary();
```
The single `&` is voluntary because I  want to unload the library even if finalisation failed.

Florent2305 replied:
Ok, maybe a comment will be useful to prevent a future bad correction.

fcollot commented on src/layers/medPython/core/medPythonCoreInit.cpp:
```diff
@@ -0,0 +1,145 @@
+                  && checkStatus(PyConfig_SetBytesString(config, &config->home, ""))
+                  && checkStatus(PyConfig_SetBytesString(config, &config->prefix, ""))
+                  && checkStatus(PyConfig_SetBytesString(config, &config->exec_prefix, ""))
+          #if PYTHON_VERSION_MINOR >= 9
```
Oui c'est défini dans le cmake.
La version la plus récente de Python est 3.9, et sur Ubuntu ils en sont encore à 3.8, donc il faut supporter au moins ces deux versions.

Florent2305 replied:
Ok merci pour les infos

fcollot commented:
If you could compile test on Windows that would help yes.
I've already compiled medPython on Ubuntu but to test it would help if someone can to compile with a fully working branch.

Florent2305 commented:
> If you could compile and test on Windows that would help yes.

I am doing it

fcollot commented on src/layers/medPython/CMakeLists.txt:
```diff
@@ -0,0 +1,194 @@
+## EXECUTABLE ##
+
+if (APPLE)
+    get_filename_component(executable_name "${Python3_EXECUTABLE}" NAME)
```
Je copie l'executable parce que je vais le modifier plus bas (ligne 112) avant de le rajouter comme ressource. La gestion des ressources de CMAKE ne permet pas de modifier le fichier après qu'il soit copié lors du packaging (du moins, je n'ai pas trouvé comment faire).

fcollot commented:
Thanks for the review @Florent2305 . I'll make commit for the remaining requested changes you made.

mathildemerle commented:
Compilation ok on Ubuntu 20.04, with `python` linked to 2.7 on my computer and `python3` to 3.8.5.

Running `medInria.sh&`, no warnings or errors, then `lsof -c medInria | grep ython` returns only `libmedPython.so` in medInria build. Seems to be ok!

fcollot commented on src/layers/medPython/core/medPythonCoreInit.cpp:
```diff
@@ -0,0 +1,145 @@
+    if (success)
+    {
+        qInfo() << "Python is initialized: " << Py_GetVersion();
+        qInfo() << getPythonPath();
```
I just realized this is a leftover debug line.

fcollot commented:
I pushed the following changes:
- Added USE_PYTHON cmake option, which controls the inclusion of medPython
- Added --no-python command line option which disables Python at execution
- Changed some of the include directives after remembering that the Python header needs to be placed first because it adds special macro definitions that affect the standard headers (on Windows).
- Removed magic number
- Improved some comments and code as suggested

fcollot commented:
While working on this I noticed it cannot possibly work on linux. @mathildemerle can you retest please and check the terminal output of the application? The output should say if Python was successfully loaded or not (it will say it failed).

fcollot commented:
(in any case there must be an output about Python or else there is an issue with the routing of the messages)

mathildemerle commented:
I'll test (i was on visio)

fcollot commented:
Just to be clear: it won't succeed because there is some code missing in the cmake part. I just need to know if it is displaying an error at execution or not.

mathildemerle commented:
Ah `WARN  - mar. juin 8 16:09:56 2021 - "Error loading the Python library:  \"Cannot load library /home/mathildemerle/Dev/medInria3.2/medInria-public/build/medInria-build/resources/medPython: (/home/mathildemerle/Dev/medInria3.2/medInria-public/build/medInria-build/resources/medPython: ne peut lire les données du fichier: est un dossier)\""`

fcollot commented:
Ok thanks.

fcollot commented:
New changes pushed. A prepackaged embeddable Python is now downloaded instead of retrieving Python from the system. (I edited the original message of this PR. Please read it for more information)
**_Please note that this will only work on your platform if the associated package has been uploaded to https://med.inria.fr/downloads/embedded-python/_. The Linux package is currently missing but should be uploaded today.**

mathildemerle commented:
Compilation from scratch on this updated branch, removing the build directory and the external projects:

```
mathildemerle@IHULUX002:~/Dev/medInria3.2/medInria-public/build$ cmake ..
-- The C compiler identification is GNU 9.3.0
-- The CXX compiler identification is GNU 9.3.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Found Boost: /usr/lib/x86_64-linux-gnu/cmake/Boost-1.71.0/BoostConfig.cmake (found version "1.71.0")  
No LSB modules are available.
No LSB modules are available.
-- Configuring done
-- Generating done
-- Build files have been written to: /home/mathildemerle/Dev/medInria3.2/medInria-public/build
```
Python not displayed here.

Then, during the compilation i've got an error:

```
/home/mathildemerle/Dev/medInria3.2/medInria-public/src/layers/medPython/core/medPythonCoreAPI.h:37:10: fatal error: Python.h: Aucun fichier ou dossier de ce type
   37 | #include <Python.h>
      |          ^~~~~~~~~~
compilation terminated.
In file included from /home/mathildemerle/Dev/medInria3.2/medInria-public/src/layers/medPython/core/medPythonCoreAPI.cpp:14:
/home/mathildemerle/Dev/medInria3.2/medInria-public/src/layers/medPython/core/medPythonCoreAPI.h:37:10: fatal error: Python.h: Aucun fichier ou dossier de ce type
   37 | #include <Python.h>
      |          ^~~~~~~~~~
compilation terminated.
make[5]: *** [layers/medPython/CMakeFiles/medPython.dir/build.make:102 : layers/medPython/CMakeFiles/medPython.dir/core/medPythonCoreInit.cpp.o] Erreur 1
make[5]: *** Attente des tâches non terminées....
make[5]: *** [layers/medPython/CMakeFiles/medPython.dir/build.make:89 : layers/medPython/CMakeFiles/medPython.dir/core/medPythonCoreAPI.cpp.o] Erreur 1
In file included from /home/mathildemerle/Dev/medInria3.2/medInria-public/src/layers/medPython/core/medPythonCoreLibrary.cpp:15:
/home/mathildemerle/Dev/medInria3.2/medInria-public/src/layers/medPython/core/medPythonCoreAPI.h:37:10: fatal error: Python.h: Aucun fichier ou dossier de ce type
   37 | #include <Python.h>
```

mathildemerle commented:
`USE_PYTHON` is at ON in medInria-build

fcollot commented:
Can you debug the cmake part to see if it's even trying to embed Python ? The code is all in the _Embed Python_ section of medPython's CMakeLists file. The python package is downloaded and extracted to the medPython build directory in the python subdirectory. You have to manually remove the directory to force a redownload.

fcollot commented:
(you have to rerun cmake on medInria after removing the python directory)

fcollot commented:
I notice you ran cmake on the superproject. Can you run it just for medInria ? I suspect it is not running cmake on the medInria project because it did not detect dependancy changes when you switched to the medPython branch.

fcollot commented:
(the dependancy issue could be fixed using a custom command later but I just need to know if the actual embedding code works)

mathildemerle commented:
Compiling from the superproject fails, but if you enter then in medInria-build, remove the `python` directory, and apply `cmake` in medInria-build the python message is displayed. However the compilation failed again in there.

mathildemerle commented:
But i did a compilation from scratch for my first run, so there were no external projets, no build, no medInria-build

fcollot commented:
I need more information to solve this issue. What was the python message that cmake displayed? When you say the compilation failed again, is it the same message as the one you posted above (missing Python.h)? What are the contents of the python directory?

mathildemerle commented:
* The same compilation error happened the second time: `#include <Python.h>`
 *  In the `python` directory there is:
 
![python](https://user-images.githubusercontent.com/3910352/122780536-37127b80-d2af-11eb-9242-51a4c28d944e.png)

 * Python.h can be found in `./layers/medPython/python/python-3.8-linux/include/python3.8/Python.h`
 * When i remove the python directory, a cmake in medInria-build does: 
```
Downloading the embeddable Python package from https://med.inria.fr/downloads/embedded-python/python-3.8-linux.zip
Download succeeded
```

mathildemerle commented:
In the CMakeList of medPython:
 * python_library_name = python3.8
 * python_working_dir = /home/mathildemerle/Dev/medInria3.2/medInria-public/build/medInria-build/layers/medPython/python
 * python_files and python_lib_files are empty (error?)

fcollot commented:
This is an issue with the zip file mechanics. It should not have unzipped the archive in a ___MACOSX/python-3.8-linux_ folder, but the contents should have been unzipped directly in the _python_ folder.

mathildemerle commented:
If i look at the linux zip on https://med.inria.fr/downloads/embedded-python/ i've got the same tree, which has been unzipped in the `python` directory. It seems to be a probleme with the zip file uploaded on the website

fcollot commented:
When I unzip on mac I do not get this problem.

fcollot commented:
We've had this sort if issue before. I'm looking into it.

fcollot commented:
solved via https://apple.stackexchange.com/questions/239578/compress-without-ds-store-and-macosx
The new files are awaiting upload. I'm looking into the other cmake issue.

mathildemerle commented:
With the new zip file, i removed my `build` directory in superproject, and compiled again in a new build dir:
 * no compilation problems
 * the new tree in `medInria-build/layers/medPython/python`:
![pythonZip](https://user-images.githubusercontent.com/3910352/122905581-a2ad2500-d351-11eb-8af4-e8a633a40a66.png)

 * The other test `qInfo() << "Python path: " << getPythonPath(); `in `medPythonCoreInit::initializePython`: i do not see the line during a `make` in medInria-build. However i've got:
```
 37%] Built target medPython_resource_python38.zip
[ 37%] Built target medPython_resource_os.py
[ 37%] Built target medPython_resource_termios.cpython-38-x86_64-linux-gnu.so
[ 37%] Built target medPython_resource_resource.cpython-38-x86_64-linux-gnu.so
[ 37%] Built target medPython_resource_readline.cpython-38-x86_64-linux-gnu.so
[ 37%] Built target medPython_resource_parser.cpython-38-x86_64-linux-gnu.so
[ 37%] Built target medPython_resource_nis.cpython-38-x86_64-linux-gnu.so
[ 37%] Built target medPython_resource_mmap.cpython-38-x86_64-linux-gnu.so
[ 37%] Built target medPython_resource_audioop.cpython-38-x86_64-linux-gnu.so
[ 37%] Built target medPython_resource__xxtestfuzz.cpython-38-x86_64-linux-gnu.so
[ 37%] Built target medPython_resource_xxlimited.cpython-38-x86_64-linux-gnu.so
[ 38%] Built target medPython_resource__testmultiphase.cpython-38-x86_64-linux-gnu.so
[ 38%] Built target medPython_resource__testinternalcapi.cpython-38-x86_64-linux-gnu.so
[ 39%] Built target medPython_resource_ossaudiodev.cpython-38-x86_64-linux-gnu.so
[ 39%] Built target medPython_resource__ssl.cpython-38-x86_64-linux-gnu.so
[ 39%] Built target medPython_resource__testbuffer.cpython-38-x86_64-linux-gnu.so
[ 39%] Built target medPython_resource__crypt.cpython-38-x86_64-linux-gnu.so
[ 39%] Built target medPython_resource__codecs_jp.cpython-38-x86_64-linux-gnu.so
[ 39%] Built target medPython_resource__contextvars.cpython-38-x86_64-linux-gnu.so
[ 39%] Built target medPython_resource__curses_panel.cpython-38-x86_64-linux-gnu.so
[ 40%] Built target medPython_resource__codecs_tw.cpython-38-x86_64-linux-gnu.so
[ 40%] Built target medPython_resource__codecs_cn.cpython-38-x86_64-linux-gnu.so
[ 40%] Built target medPython_resource__sqlite3.cpython-38-x86_64-linux-gnu.so
[ 40%] Built target medPython_resource__bz2.cpython-38-x86_64-linux-gnu.so
[ 40%] Built target medPython_resource__ctypes_test.cpython-38-x86_64-linux-gnu.so
[ 40%] Built target medPython_resource__uuid.cpython-38-x86_64-linux-gnu.so
[ 40%] Built target medPython_resource__hashlib.cpython-38-x86_64-linux-gnu.so
[ 40%] Built target medPython_resource__codecs_iso2022.cpython-38-x86_64-linux-gnu.so
[ 40%] Built target medPython_resource__curses.cpython-38-x86_64-linux-gnu.so
[ 40%] Built target medPython_resource__testimportmultiple.cpython-38-x86_64-linux-gnu.so
[ 40%] Built target medPython_resource__ctypes.cpython-38-x86_64-linux-gnu.so
[ 40%] Built target medPython_resource_python3.8
[ 40%] Built target medPython_resource__decimal.cpython-38-x86_64-linux-gnu.so
[ 40%] Built target medPython_resource__codecs_hk.cpython-38-x86_64-linux-gnu.so
[ 41%] Built target medPython_resource__gdbm.cpython-38-x86_64-linux-gnu.so
[ 42%] Built target medPython_resource__posixshmem.cpython-38-x86_64-linux-gnu.so
[ 42%] Built target medPython_resource__json.cpython-38-x86_64-linux-gnu.so
[ 42%] Built target medPython_resource__lsprof.cpython-38-x86_64-linux-gnu.so
[ 42%] Built target medPython_resource__asyncio.cpython-38-x86_64-linux-gnu.so
[ 42%] Built target medPython_resource__lzma.cpython-38-x86_64-linux-gnu.so
[ 42%] Built target medPython_resource__codecs_kr.cpython-38-x86_64-linux-gnu.so
[ 42%] Built target medPython_resource__dbm.cpython-38-x86_64-linux-gnu.so
[ 42%] Built target medPython_resource__multibytecodec.cpython-38-x86_64-linux-gnu.so
[ 43%] Built target medPython_resource_LICENSE.txt
[ 43%] Built target medPython_resource__multiprocessing.cpython-38-x86_64-linux-gnu.so
[ 43%] Built target medPython_resource__xxsubinterpreters.cpython-38-x86_64-linux-gnu.so
[ 43%] Built target medPython_resource__testcapi.cpython-38-x86_64-linux-gnu.so
[ 43%] Automatic MOC for target medPython
[ 43%] Built target medPython_autogen
Scanning dependencies of target medPython
[the warning shown next point]
[ 43%] Linking CXX shared library ../../lib/libmedPython.so
[ 44%] Built target medPython
[ 44%] Built target medPython_resource__opcode.cpython-38-x86_64-linux-gnu.so
[ 44%] Built target medPython_resource__queue.cpython-38-x86_64-linux-gnu.so
```

 * By the way, a warning:
```
/home/mathildemerle/Dev/medInria3.2/medInria-public/src/layers/medPython/core/medPythonCoreInit.cpp: In function ‘bool med::python::core::{anonymous}::prepareConfig(PyConfig*)’:
/home/mathildemerle/Dev/medInria3.2/medInria-public/src/layers/medPython/core/medPythonCoreInit.cpp:98:27: warning: comparison of integer expressions of different signedness: ‘int’ and ‘const size_t’ {aka ‘const long unsigned int’} [-Wsign-compare]
   98 |         for (int i = 0; i < numModuleSearchPaths; i++)
```

fcollot commented:
Thanks for the tests (the line I asked to add is to display the path at runtime)
I'm making changes to the cmake file to handle the dependancy problem and will push them soon. Also, I updated the branch with the changes that were made to `add_external_resources`, so the "built target' lines will no longer appear.

mathildemerle commented:
(indeed for the qInfo line sorry XD)

Running medInria:
```
WARN  - mar. juin 22 12:45:57 2021 - "Error loading the Python library:  \"Cannot load library /home/mathildemerle/Dev/medInria3.2/medInria-public/build/medInria-build/resources/medPython/python3.8: (/home/mathildemerle/Dev/medInria3.2/medInria-public/build/medInria-build/resources/medPython/python3.8: ne peut dynamiquement charger un exécutable)\""
```

fcollot commented:
I've completely refactored this PR. The executable is not used anymore (because it's not possible on linux) so we now load the (correct) shared library on all platforms. Without going into the details (because I didn't quite understand it all), this required some changes. Python could no longer load some modules from the ZIP file, so I had to move them out. This was especially problematic for the _encodings_ module because it is a package, and the resource system does not accept directories. I hacked a solution by creating a "virtual" package at runtime (which consists of a temporary directory containing symbolic links to the encoding files in the resource folder). Also, you'll notice the Python files are all listed in the cmake project. This is necessary to properly handle dependencies (and it's not complicated to generate).
The new embeddable python archives can be downloaded at https://we.tl/t-DOwoLlSHsR. This allows you to test without waiting for the files to be uploaded to the medInria website. All you have to do is copy the archive that corresponds to your platform to the python subdirectory of medPython's build directory, and rename it to _python.zip_.

mathildemerle commented on src/layers/medPython/CMakeLists.txt:
```diff
@@ -57,76 +64,72 @@ target_include_directories(${TARGET_NAME}
-foreach (python_file ${python_files})
-    add_external_resources(${TARGET_NAME} "${python_file}")
-endforeach()
+# To prevent warniong of generated files not being moc'd
```
*warning

mathildemerle commented:
Running the new version with the zip i added manually:
```
INFO  - lun. juil. 5 11:42:30 2021 - "Setting up Python..." 
INFO  - lun. juil. 5 11:42:30 2021 - "Successfully loaded the Python library:  \"/home/mathildemerle/Dev/medInria3.2/medInria-public/build/medInria-build/resources/medPython/libpython3.8.so\"" 
INFO  - lun. juil. 5 11:42:30 2021 - "Resolving CPython symbols" 
INFO  - lun. juil. 5 11:42:30 2021 - "Python is initialized:  3.8.5 (default, Jan 27 2021, 15:41:15) \n[GCC 9.3.0]"
```

mathildemerle commented on src/layers/medPython/core/medPythonCoreInit.cpp:
```diff
@@ -0,0 +1,291 @@
+namespace med::python::core
+{
+
+namespace
```
is this necessary?

fcollot replied:
For a simple answer: https://google.github.io/styleguide/cppguide.html#Internal_Linkage
(I use namespaces instead of singleton classes, and elements inside unnamed namespaces is sort of the equivalent of private members for a class)

mathildemerle commented on src/layers/medPython/core/medPythonCoreLibrary.cpp:
```diff
@@ -0,0 +1,131 @@
+namespace med::python::core
+{
+
+namespace
```
is this needed also?
